container users and groups from process root

deepskyblue86 opened the pull request:
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area libsinsp

/area tests

**Does this PR require a change in the driver versions?**

**What this PR does / why we need it**:
Introduce `add_container_user` and `add_container_group`.
Those are called by `sinsp_threadinfo` `set_user` / `set_group` and load the information from /proc/<pid>/root/etc.
For that to work, the process running sinsp has to be privileged and in the host pid namespace.
For more information see `man 5 proc`.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

FedeDP commented on userspace/libsinsp/user.cpp:
```diff
@@ -251,51 +284,66 @@ scap_userinfo *sinsp_usergroup_manager::add_user(const string &container_id, uin
+	scap_userinfo *retval{nullptr};
+
+#if defined HAVE_PWD_H && defined HAVE_FGET__ENT
+	std::string proc_root_etc = s_host_root + "/proc/" + std::to_string(pid) + "/root/etc/";
```
This is a freakingly neat idea! Thanks for this!
@terylt @araujof

deepskyblue86 replied:
Credits to @gnosek for teaching me that. Credits to myself for a few days of experiments :rofl:

loresuso replied:
Really nice!

FedeDP reviewed (commented):
I love these changes!
I am going to give this a bit of testing asap and come back to you!

FedeDP commented:
Also, the CI did explode :D 
btw @loresuso you might be interested in this since you developed the old solution with me!

Finally, i think we can also drop the container engines part to load the `lowerdir` .

deepskyblue86 commented:
Yeah, the UT (which I actually wrote myself :sweat_smile:) failed, I'll update this PR soon!

deepskyblue86 commented:
> Also, the CI did explode :D btw @loresuso you might be interested in this since you developed the old solution with me!
> 
> Finally, i think we can also drop the container engines part to load the `lowerdir` .

@FedeDP do you want me to drop that in this PR?

FedeDP commented:
Yep, thanks!

deepskyblue86 commented:
I found something weird: if I checkout `0.9.0` `sinsp-example` behaves as expected, while on `master` I have `{"error": ""}` and no container events (I'm doing `-j -f 'evt.type in (execve, execveat) and evt.dir=< and container.id != host'`).

Molter73 commented:
I rebased the PR for adding e2e tests to CI on Tuesday, so whatever broke them is a change after that. I'll try to look into them ASAP, but if you guys find what is wrong, you can open a PR.

FedeDP commented:
Hi Mauro! I think the fix is #678 ! Thanks to Angelo <3

FedeDP commented:
We can rebase on top of master :)

FedeDP reviewed (dismissed):
/approve
I really like this :) 
We can cleanup the overalyfs things later, in a subsequent PR!

poiana commented:
LGTM label has been added.  <details>Git tree hash: 7afc1b7e23321ddd423db20bd77bfe3181c566e3</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/677#" title="Author self-approved">deepskyblue86</a>*, *<a href="https://github.com/falcosecurity/libs/pull/677#pullrequestreview-1150611179" title="Approved">FedeDP</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP reviewed (commented):
Mmh on a better look, it seems like the container start is triggering an useradded event for host users (tested with sinsp-example: 
```
sudo ./libsinsp/examples/sinsp-example --bpf driver/bpf/probe.o -f evt.type=useradded -o "%evt.arg[2] %evt.arg[5]"
```
* evt.arg[2] is user name
* evt.arg[5] is container id


It prints:
```
{"evt.arg[2]":"root","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"bin","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"daemon","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"mail","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"ftp","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"http","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"nobody","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"dbus","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"systemd-journal-remote","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"systemd-network","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"systemd-oom","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"systemd-resolve","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"systemd-timesync","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"systemd-coredump","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"uuidd","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"avahi","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"polkitd","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"rtkit","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"sddm","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"usbmux","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"federico","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"tss","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"git","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"geoclue","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"nm-openvpn","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"openvpn","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"brltty","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"sophosav","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"redis","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"dnsmasq","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"libvirt-qemu","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"mysql","evt.arg[5]":"2d4a2dd77f70"}
{"evt.arg[2]":"qemu","evt.arg[5]":"2d4a2dd77f70"}
```
Note the "federico" in there; these are the host users.

FedeDP commented:
/hold

poiana commented:
New changes are detected. LGTM label has been removed.

deepskyblue86 commented:
> Mmh on a better look, it seems like the container start is triggering an useradded event for host users (tested with sinsp-example:
> 
> ```
> sudo ./libsinsp/examples/sinsp-example --bpf driver/bpf/probe.o -f evt.type=useradded -o "%evt.arg[2] %evt.arg[5]"
> ```
> 
> * evt.arg[2] is user name
> * evt.arg[5] is container id
> 
> It prints:
> 
> ```
> {"evt.arg[2]":"root","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"bin","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"daemon","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"mail","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"ftp","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"http","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"nobody","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"dbus","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"systemd-journal-remote","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"systemd-network","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"systemd-oom","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"systemd-resolve","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"systemd-timesync","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"systemd-coredump","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"uuidd","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"avahi","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"polkitd","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"rtkit","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"sddm","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"usbmux","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"federico","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"tss","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"git","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"geoclue","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"nm-openvpn","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"openvpn","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"brltty","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"sophosav","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"redis","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"dnsmasq","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"libvirt-qemu","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"mysql","evt.arg[5]":"2d4a2dd77f70"}
> {"evt.arg[2]":"qemu","evt.arg[5]":"2d4a2dd77f70"}
> ```
> 
> Note the "federico" in there; these are the host users.

@FedeDP I fixed that / changed the approach a bit. I also updated the e2e test, but unfortunately I can't test locally.

terylt commented:
Nice work! This looks like it'll be a simpler approach to user management for containers.   Will try to test in the next couple of weeks.

deepskyblue86 commented:
Warning: this approach doesn't work with fuse-overlayfs (I saw that on podman)

FedeDP commented on userspace/libsinsp/user.cpp:
```diff
@@ -32,7 +32,88 @@ limitations under the License.
+
+const char *get_root_ns_mnt()
+{
+	static char *s_root_ns_mnt = nullptr;
```
Can we make this a class field? And free it in the dtor?
Same for `CANNOT_READ_ROOT_NS_MNT`.
Perhaps we can load them in the ctor.

deepskyblue86 replied:
In its own class!

FedeDP commented on userspace/libsinsp/user.cpp:
```diff
@@ -243,59 +357,110 @@ scap_userinfo *sinsp_usergroup_manager::add_user(const string &container_id, uin
-#ifdef HAVE_PWD_H
-		struct passwd *p = nullptr;
-		if (container_id.empty() && !name)
+		bool do_notify{false};
```
I'd renamed this bool to `usermap_updated` or something more meaningful!

deepskyblue86 replied:
`inserted`

FedeDP commented on userspace/libsinsp/user.cpp:
```diff
@@ -328,35 +493,31 @@ scap_groupinfo *sinsp_usergroup_manager::add_group(const string &container_id, u
-#ifdef HAVE_GRP_H
-		struct group *p = nullptr;
-		if (container_id.empty() && !name)
+		bool do_notify{false};
```
Same as above ;)

FedeDP reviewed (commented):
Overall this LGTM; i left some minor comment!
Thanks Angelo!

mstemm commented:
Since these lookups are being done live, how are they going to work with trace files? Generally any info that ends up in the reconstructed state has to either be in the initial process/user/container list fetched by libscap, or you need an event that injects the new info.

Is this info added to the useradded event?

FedeDP commented on userspace/libsinsp/procfs_utils.cpp:
```diff
@@ -63,3 +65,60 @@ std::string libsinsp::procfs_utils::get_systemd_cgroup(std::istream& cgroups)
+		return false;
+	}
+
+	if(0 == strncmp(m_host_init_ns_mnt, proc_ns_mnt, NS_MNT_SIZE))
```
Why not using just `strcmp` here?

deepskyblue86 replied:
Just habit of using the bounded version, no specific reason.

FedeDP commented on userspace/libsinsp/procfs_utils.h:
```diff
@@ -27,5 +27,43 @@ int get_userns_root_uid(std::istream& uid_map);
+
+private:
+	const std::string& m_host_root;
+	char* m_host_init_ns_mnt{nullptr};
```
Using a unique ptr?

deepskyblue86 replied:
That's being used with C functions, I just felt it natural to keep it C-style :)

deepskyblue86 commented:
> Since these lookups are being done live, how are they going to work with trace files? Generally any info that ends up in the reconstructed state has to either be in the initial process/user/container list fetched by libscap, or you need an event that injects the new info.
> 
> Is this info added to the useradded event?

It was already done with `notify_user_changed` / `notify_group_changed` which I used too.

FedeDP reviewed (dismissed):
/approve

poiana commented:
LGTM label has been added.  <details>Git tree hash: 0c885bd7823f07e6a0b554b611ec975b279f3b46</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/677#" title="Author self-approved">deepskyblue86</a>*, *<a href="https://github.com/falcosecurity/libs/pull/677#pullrequestreview-1171656607" title="Approved">FedeDP</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
New changes are detected. LGTM label has been removed.

deepskyblue86 commented:
@Molter73 I'd like you to check the e2e part, if you have some time. Thanks in advance.

Molter73 commented on userspace/libsinsp/procfs_utils.cpp:
```diff
@@ -63,3 +65,60 @@ std::string libsinsp::procfs_utils::get_systemd_cgroup(std::istream& cgroups)
+	std::string path = m_host_root + "/proc/" + std::to_string(pid) + "/ns/mnt";
+
+	char proc_ns_mnt[NS_MNT_SIZE] = {0};
+	if(-1 == readlink(path.c_str(), proc_ns_mnt, NS_MNT_SIZE))
```
Similar to my previous comment on `readlink`, though not as serious since it is used in `strncmp` later.
```suggestion
	if(-1 == readlink(path.c_str(), proc_ns_mnt, NS_MNT_SIZE - 1))
```

Molter73 commented on userspace/libsinsp/user.cpp:
```diff
@@ -243,59 +282,110 @@ scap_userinfo *sinsp_usergroup_manager::add_user(const string &container_id, uin
+
+	if(false == m_ns_helper->in_own_ns_mnt(pid))
+	{
+		return nullptr;
```
Might as well use it
```suggestion
		return retval;
```

deepskyblue86 replied:
I choose to use `nullptr` for readability, to highlight we're not returning a value

deepskyblue86 commented on userspace/libsinsp/procfs_utils.cpp:
```diff
@@ -63,3 +65,60 @@ std::string libsinsp::procfs_utils::get_systemd_cgroup(std::istream& cgroups)
+{
+	// (try to) init m_host_init_ns_mnt
+	char buf[NS_MNT_SIZE] = {0};
+	if(-1 == readlink((m_host_root + "/proc/1/ns/mnt").c_str(), buf, NS_MNT_SIZE))
```
True, even if `NS_MNT_SIZE` is really big :grin:

deepskyblue86 commented on userspace/libsinsp/procfs_utils.cpp:
```diff
@@ -63,3 +65,60 @@ std::string libsinsp::procfs_utils::get_systemd_cgroup(std::istream& cgroups)
+	char buf[NS_MNT_SIZE] = {0};
+	if(-1 == readlink((m_host_root + "/proc/1/ns/mnt").c_str(), buf, NS_MNT_SIZE))
+	{
+		g_logger.format(sinsp_logger::SEV_WARNING,
+				"Cannot read host init ns/mnt");
```
AFAIK the only reason would be permission denied, but ok

Molter73 reviewed (approved):
/lgtm

poiana commented:
LGTM label has been added.  <details>Git tree hash: 60da174f5558ca010ad752432af9a5eb3ecf1ca1</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/677#" title="Author self-approved">deepskyblue86</a>*, *<a href="https://github.com/falcosecurity/libs/pull/677#pullrequestreview-1171656607" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/677#pullrequestreview-1187988515" title="LGTM">Molter73</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [FedeDP,Molter73]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

gnosek commented on userspace/libsinsp/threadinfo.cpp:
```diff
@@ -505,8 +505,21 @@ void sinsp_threadinfo::set_user(uint32_t uid)
+		{
+			user = m_inspector->m_usergroup_manager.add_user(m_container_id, uid, m_group.gid, NULL, NULL, NULL, m_inspector->is_live());
+		}
+		else if(uid != 0)
+		{
+			//
+			// When a container is running with a specific user and this
+			// get called with 0, it's too early to make an attempt.
+			// As a downside we won't load users for containers running as
+			// root, but we will load them if e.g.docker exec -u <specific-user>.
+			//
+			user = m_inspector->m_usergroup_manager.add_container_user(m_container_id, m_pid, uid, m_inspector->is_live());
+		}
```
Why is it "too early"? Because we caught a thread switching into the container? What if the container _is_ started as root?

What *is* `uid` here? If we're calling this from all sorts of parsers, it's not a special uid, like the `docker -u` user, right? So how does the `uid != 0` check work?

Maybe hardcoding uid=0 as "root" would be good enough btw?

FedeDP replied:
@deepskyblue86 :)

deepskyblue86 replied:
1. Why is it "too early"?
  By trial and error, I found that the thread had the container id but wasn't really _inside_ the container (i.e. `runc:[1:CHILD]` in my tests).
2. What if the container is started as root?
  We'll have `<NA>` :shrug: but if / when it switches to a different user we'll load the users.
3. What is `uid` here?
  a) `sinsp_threadinfo::set_user` is being called by `parse_chroot_exit`, `parse_clone_exit`, `parse_execve_exit`, `parse_setuid_exit`, `parse_setresuid_exit`.
  b) It's actually the user of `docker -u` or k8s `runAsUser`.
  c) you can see that it's being tested (test/e2e/tests/test_process/test_container.py) by starting a `hashicorp/http-echo:alpine` container with `11:110`
4. Maybe hardcoding uid=0 as "root" would be good enough btw?
  I was a bit uncertain about doing so, but if it's already two of us I'd go with that :grinning:

FedeDP replied:
> Maybe hardcoding uid=0 as "root" would be good enough btw?

Makes sense.

gnosek reviewed (commented):
Just one question, LGTM otherwise :)

FedeDP reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/677#" title="Author self-approved">deepskyblue86</a>*, *<a href="https://github.com/falcosecurity/libs/pull/677#pullrequestreview-1189592019" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/677#pullrequestreview-1187988515" title="LGTM">Molter73</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [FedeDP,Molter73]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP commented:
Sorry @gnosek i might have left behind your comment and approved :( 
Hopefully i was able to hold this.

FedeDP commented:
Uh i already put this on hold 20d ago ahah @deepskyblue86 care to answer to grzeg question?

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/677#" title="Author self-approved">deepskyblue86</a>*, *<a href="https://github.com/falcosecurity/libs/pull/677#pullrequestreview-1189592019" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/677#pullrequestreview-1192083603" title="Approved">gnosek</a>*, *<a href="https://github.com/falcosecurity/libs/pull/677#pullrequestreview-1187988515" title="LGTM">Molter73</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [FedeDP,Molter73,gnosek]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP commented:
/unhold
